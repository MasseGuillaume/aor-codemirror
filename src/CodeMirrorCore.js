import React from 'react'
import PropTypes from 'prop-types'
import { Controlled as ControlledCodeMirror } from 'react-codemirror2'
import { Button, Toolbar } from '@material-ui/core'
import { editorTypes } from './editorTypes'

import 'codemirror/lib/codemirror.css'
import 'codemirror/mode/xml/xml'
import 'codemirror/mode/sql/sql'
import 'codemirror/mode/javascript/javascript'

export const CodeMirrorCore = ({
  input: { value, onChange = () => {} } = {},
  meta: { valid, error },
  type,
  label,
  defaultValue = '',
  required = false
}) => {
  const { mode, format, parse, stringify } = editorTypes[type]
  return (
    <div>
      <span style={{ color: '#c4c4c4' }}>
        {label}
        {required ? ' * ' : ' '}
      </span>
      {valid ? null : <span style={{ color: '#ff3d3a' }}> ({error})</span>}
      {Boolean(format) && (
        <Toolbar disableGutters>
          <Button onClick={() => onChange(format(value))}>Format</Button>
        </Toolbar>
      )}
      <ControlledCodeMirror
        label={label}
        value={stringify(value || defaultValue)}
        options={{ mode, lineNumbers: true, lineWrapping: true }}
        onBeforeChange={(editor, data, valueFromEditor) => {
          onChange(parse(valueFromEditor))
        }}
      />
    </div>
  )
}
const valueType = PropTypes.oneOfType([PropTypes.string, PropTypes.array, PropTypes.object])
CodeMirrorCore.propTypes = {
  input: PropTypes.shape({
    value: valueType,
    onChange: PropTypes.func
  }),
  meta: PropTypes.shape({
    valid: PropTypes.bool.isRequired,
    error: PropTypes.string
  }).isRequired,
  type: PropTypes.oneOf(Object.keys(editorTypes)).isRequired,
  label: PropTypes.node.isRequired,
  defaultValue: valueType,
  required: PropTypes.bool
}
