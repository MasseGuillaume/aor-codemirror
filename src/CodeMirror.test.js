import { CodeMirror } from './CodeMirror'
import { CodeMirrorCore } from './CodeMirrorCore'

const defaultProps = {
  name: 'fieldName',
  label: 'Field Label',
  type: 'json'
}

describe('CodeMirror', () => {
  it('renders with defaults', () => {
    const wrapper = shallow(<CodeMirror {...defaultProps} />)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper).toHaveProp('component', CodeMirrorCore)
  })

  it('renders with source', () => {
    expect(shallow(<CodeMirror {...defaultProps} source="sourceObject" />)).toMatchSnapshot()
  })
})
