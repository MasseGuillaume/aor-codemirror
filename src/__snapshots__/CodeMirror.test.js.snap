// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeMirror renders with defaults 1`] = `
<Field
  component={[Function]}
  label="Field Label"
  name="fieldName"
  type="json"
/>
`;

exports[`CodeMirror renders with source 1`] = `
<Field
  component={[Function]}
  label="Field Label"
  name="sourceObject.fieldName"
  source="sourceObject"
  type="json"
/>
`;
